pkg/config/mesh/meshwatcher/mesh.go

```diff
@@ -0,0 +1,94 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package meshwatcher
+
+import (
+	uatomic "go.uber.org/atomic"
+	"google.golang.org/protobuf/proto"
+
+	meshconfig "istio.io/api/mesh/v1alpha1"
+	"istio.io/istio/pkg/config/mesh"
+	"istio.io/istio/pkg/kube/krt"
+	"istio.io/istio/pkg/util/protomarshal"
+)
+
+// WatcherCollection is an interface to describe an object that implements both the legacy mesh.Watcher interface and the
+// new krt interface.
+type WatcherCollection interface {
+	mesh.Watcher
+	krt.Singleton[MeshConfigResource]
+}
+
+func ConfigAdapter(configuration krt.Singleton[MeshConfigResource]) WatcherCollection {
+	return adapter{configuration}
+}
+
+type adapter struct {
+	krt.Singleton[MeshConfigResource]
+}
+
+func (a adapter) Mesh() *meshconfig.MeshConfig {
+	v := a.Singleton.Get()
+	return v.MeshConfig
+}
+
+func (a adapter) AddMeshHandler(h func()) *mesh.WatcherHandlerRegistration {
+	active := uatomic.NewBool(true)
+	reg := mesh.NewWatcherHandlerRegistration(func() {
+		active.Store(false)
```

Comment:
I'm confused; we say we're no longer active when we get an event?

Reply:
Ah this is a bit confusing -- the function we pass is the `remove` callback. So when we remove this is called.

I am now seeing I never finished my "TODO: go back and comment things" ...